usb/ehci/ehci.c

```diff
@@ -0,0 +1,737 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * ehci/ehci.c
+ *
+ * Copyright 2018, 2021 Phoenix Systems
+ * Copyright 2007 Pawel Pisarczyk
+ * Author: Jan Sikorski, Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <sys/mman.h>
+#include <sys/threads.h>
+#include <sys/list.h>
+#include <sys/interrupt.h>
+#include <sys/minmax.h>
+
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <usbhost.h>
+#include <stdio.h>
+
+#include <hub.h>
+#include <hcd.h>
+
+#include "ehci.h"
+
+#define EHCI_PERIODIC_SIZE 128
+
+struct qtd_node {
+	struct qtd_node *prev, *next;
+	struct qtd *qtd;
+};
+
+
+struct qh_node {
+	struct qh_node *prev, *next;
+	struct qtd *last;
+	struct qh *qh;
+	unsigned period; /* [ms], interrupt transfer only */
+	unsigned phase; /* [ms], interrupt transfer only */
+	unsigned uframe; /* interrupt transfer and high-speed only */
+};
+
+
+static int ehci_linkQtd(struct qtd *prev, struct qtd *next)
+{
+	prev->next.pointer = va2pa(next) >> 5;
+	prev->next.type = 0;
+	prev->next.terminate = 0;
+
+	return 0;
+}
+
+
+static void ehci_enqueue(struct qh_node *qhnode, struct qtd *first, struct qtd *last)
+{
+	struct qtd *prevlast = qhnode->last;
+
+	qhnode->last = last;
+	last->next.terminate = 1;
+	last->ioc = 1;
+
+	prevlast->next.pointer = va2pa(first) >> 5;
+	prevlast->next.type = 0;
+
+	asm volatile ("dmb" ::: "memory");
+
```

Comment:
It's best to separate asm from C. Maybe static inline wrapper for memory barier?